spring-integration-core/src/main/java/org/springframework/integration/util/MessagingMethodInvokerHelper.java

```diff
@@ -513,43 +515,74 @@ private void initializeHandler(HandlerMethod candidate) {
-
-			this.messageHandlerMethodFactory.setCustomArgumentResolvers(customArgumentResolvers);
-
-			if (getBeanFactory() != null &&
-					getBeanFactory()
-							.containsBean(IntegrationContextUtils.ARGUMENT_RESOLVER_MESSAGE_CONVERTER_BEAN_NAME)) {
-				this.messageHandlerMethodFactory
-						.setMessageConverter(getBeanFactory()
-								.getBean(IntegrationContextUtils.ARGUMENT_RESOLVER_MESSAGE_CONVERTER_BEAN_NAME,
-										MessageConverter.class));
+			if (getBeanFactory() != null
+					&& getBeanFactory().containsBean(
+							IntegrationContextUtils.ARGUMENT_RESOLVER_MESSAGE_CONVERTER_BEAN_NAME)) {
+				try {
+					this.messageHandlerMethodFactory.setMessageConverter(getBeanFactory().getBean(
+							IntegrationContextUtils.ARGUMENT_RESOLVER_MESSAGE_CONVERTER_BEAN_NAME,
+							MessageConverter.class));
+					if (this.canProcessMessageList) {
+						this.messageHandlerMethodFactory.setCustomArgumentResolvers(getBeanFactory().getBean(
+								IntegrationContextUtils.LIST_ARGUMENT_RESOLVERS_BEAN_NAME,
+								HandlerMethodArgumentResolversHolder.class).getResolvers());
+					}
+					else {
+						this.messageHandlerMethodFactory.setCustomArgumentResolvers(getBeanFactory().getBean(
+								IntegrationContextUtils.ARGUMENT_RESOLVERS_BEAN_NAME,
+								HandlerMethodArgumentResolversHolder.class).getResolvers());
+					}
+				}
+				catch (NoSuchBeanDefinitionException e) {
+					configureLocalMessageHandlerFactory();
+				}
+			}
+			else {
+				configureLocalMessageHandlerFactory();
 			}
-
 			this.messageHandlerMethodFactory.afterPropertiesSet();
 			prepareEvaluationContext();
 			this.initialized = true;
 		}
 	}
 
+	/*
+	 * This should not be needed in production but we have many tests
+	 * that don't run in an application context.
+	 */
+	private void configureLocalMessageHandlerFactory() {
+		PayloadExpressionArgumentResolver payloadExpressionArgumentResolver =
+				new PayloadExpressionArgumentResolver();
+		payloadExpressionArgumentResolver.setBeanFactory(getBeanFactory());
+
+		PayloadsArgumentResolver payloadsArgumentResolver = new PayloadsArgumentResolver();
+		payloadsArgumentResolver.setBeanFactory(getBeanFactory());
+
+		CollectionArgumentResolver collectionArgumentResolver =
+				new CollectionArgumentResolver(this.canProcessMessageList);
+		collectionArgumentResolver.setBeanFactory(getBeanFactory());
+
+		MapArgumentResolver mapArgumentResolver = new MapArgumentResolver();
+		mapArgumentResolver.setBeanFactory(getBeanFactory());
+
+		List<HandlerMethodArgumentResolver> customArgumentResolvers = new LinkedList<>();
+		customArgumentResolvers.add(payloadExpressionArgumentResolver);
+		customArgumentResolvers.add(payloadsArgumentResolver);
+		customArgumentResolvers.add(collectionArgumentResolver);
+		customArgumentResolvers.add(mapArgumentResolver);
+
+		this.messageHandlerMethodFactory.setCustomArgumentResolvers(customArgumentResolvers);
+
+		if (getBeanFactory() != null &&
```

Comment:
Redundant code. If we are in this `configureLocalMessageHandlerFactory()`, then no `beanFactory` or `ARGUMENT_RESOLVER_MESSAGE_CONVERTER_BEAN_NAME`. Therefore this code is dead.
Will fix on merge

Reply:
Not really - this is called from 2 places - bf exists but doesn't have the arg resolver beans.

Reply:
Yeah... Makes sense. Thank you!